### PR TITLE
rgw-admin: new command line parsing

### PR DESCRIPTION
**Pull request by yehudasa**
New command line parsing, gets rid of old manual code. Also dump list of expected commands in case command is not found.

I went over the different commands and made sure that all the previous commands would still work. There are a bunch of hyphenated commands that in some cases I added a non hyphenated version of the same command.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -395,6 +395,7 @@ void usage()
   generic_client_usage();
 }
 
+
```
Comment (mattbenjamin) [inline src/rgw/rgw_admin.cc]: I like the minimalism.  Maybe we could make this an enum class while we're here?

File: src/rgw/rgw_admin.cc [hunk 2]
```diff
@@ -515,365 +515,372 @@ class SimpleCmd {
...
-  OPT_PUBSUB_TOPIC_RM,
-  OPT_PUBSUB_NOTIFICATION_CREATE,
-  OPT_PUBSUB_NOTIFICATION_RM,
-  OPT_PUBSUB_SUB_GET,
-  OPT_PUBSUB_SUB_CREATE,
-  OPT_PUBSUB_SUB_RM,
-  OPT_PUBSUB_SUB_PULL,
-  OPT_PUBSUB_EVENT_RM,
+
+namespace rgw_admin {
```
Comment (mattbenjamin) [inline src/rgw/rgw_admin.cc]: yes, this is superb

File: src/rgw/rgw_admin.cc [hunk 3]
```diff
@@ -515,365 +515,372 @@ class SimpleCmd {
...
-  { "pubsub topic get", OPT_PUBSUB_TOPIC_GET },
-  { "pubsub topic rm", OPT_PUBSUB_TOPIC_RM },
-  { "pubsub notification create", OPT_PUBSUB_NOTIFICATION_CREATE },
-  { "pubsub notification rm", OPT_PUBSUB_NOTIFICATION_RM },
-  { "pubsub sub get", OPT_PUBSUB_SUB_GET },
-  { "pubsub sub create", OPT_PUBSUB_SUB_CREATE },
-  { "pubsub sub rm", OPT_PUBSUB_SUB_RM },
-  { "pubsub sub pull", OPT_PUBSUB_SUB_PULL },
-  { "pubsub event rm", OPT_PUBSUB_EVENT_RM },
+  { "user create", OPT::USER_CREATE },
```
Comment (mattbenjamin) [inline src/rgw/rgw_admin.cc]: namespace ftw

File: src/rgw/rgw_admin.cc [hunk 4]
```diff
@@ -400,7 +400,7 @@ class SimpleCmd {
 public:
   struct Def {
     string cmd;
-    int opt{-1};
+    std::any opt;
```
Comment (mattbenjamin) [inline src/rgw/rgw_admin.cc]: yes, that's an improvement

File: src/rgw/rgw_admin.cc [hunk 5]
```diff
@@ -396,6 +396,125 @@ void usage()
 }
 
 
+class SimpleCmd {
```
Comment (mattbenjamin) [inline src/rgw/rgw_admin.cc]: seems like a good idea

File: src/rgw/rgw_admin.cc [hunk 6]
```diff
@@ -771,6 +773,7 @@ static SimpleCmd::Commands all_cmds = {
   { "orphans find", OPT::ORPHANS_FIND },
   { "orphans finish", OPT::ORPHANS_FINISH },
   { "orphans list jobs", OPT::ORPHANS_LIST_JOBS },
+  { "orphans list-jobs", OPT::ORPHANS_LIST_JOBS },
```
Comment (mattbenjamin) [inline src/rgw/rgw_admin.cc]: ok, buy this

**Comment by ivancich** [comment]
@yehudasa Will this need a tracker, so we can track backports?

**Comment by yehudasa** [comment]
@ivancich created: https://tracker.ceph.com/issues/42513

**Comment by cbodley** [comment]
is there anything in this pr that isn't in https://github.com/ceph/ceph/pull/31686? if not, lets close this. otherwise needs a rebase

**Comment by yehudasa** [comment]
part of https://github.com/ceph/ceph/pull/31686

**Pull request closed without merging**